### PR TITLE
Support changing labels of Actions runner without re-registration

### LINKED ISSUE #24540: [Proposal] Support changing labels of Actions runner without re-registration
## Background

Act runner will ask for labels when you register it to a Gitea instance. For example, one of the default labels is `ubuntu-latest:docker://node:16-bullseye` . It could be split into two parts.:

- `ubuntu-latest`: This means it can run jobs with `runs-on: ubuntu-latest`
- `docker://node:16-bullseye`: This means that the jobs will run in Docker containers with the specified image.

The runner will report `ubuntu-latest` to the Gitea instance, so the instance will know what kind of jobs can be assigned to this runner. However, the instance doesn't know the second part of the label because Gitea doesn't care how a runner runs jobs.

For some reason, it was designed to allow modifying labels on the Gitea side, in a special way:

- Gitea cannot modify the labels reported by runners, which are called "Agent Labels".
- Gitea can append more labels to runners, which are called "Custom Labels".

Untitled

So it will be convenient for users when they have a job with `runs-on: centos`, but there are no runners with the label. They can add `centos` as a custom label to let the runner pick up the jobs.

Unfortunately, when the runner receives the job, it doesn't understand how to run it. To avoid failure, the act runner will use the default image to run jobs with unrecognized labels. This may surprise users.

IMO, the point is that we shouldn't let Gitea define runners. It should be the runners who decide what kind of jobs they can run, not the Gitea instance. However, we also need to provide an elegant way to modify labels on the runner side, instead of re-registration.

### Forbid modifying runner labels on the Gitea side

1. ~Combine “agent labels” and “custom labels” to “labels”.~ Drop “custom labels” and use “agent labels” only.
2. Users cannot modify the labels on settings pages of Gitea.

### Provide a new RPC method for runners to declare their labels

1. A new method like `Declare`(or a better name).
2. Act runner will call `Declare` to pass its current labels to Gitea once it has started.
3. Gitea will update the labels of the runner in database.

### Support specifying labels in the config file of the runner

1. Users can specify labels in the config file; it's optional. An empty value means "none" instead of "empty”.
2. When registering a runner:
    1. Ignore the step/flag for labels if a config file with labels has been specified.
    2. Otherwise, ask for labels.
3. When running a runner:
    1. Overwrite the labels to state file (.runner) if a config file with labels has been specified, and declare with new labels to Gitea.
    2. Otherwise, use the labels stored in the state file.

## Benefits

1. Provide convenience. Users can add/remove/change the labels of a runner by simply updating the config file and restarting the runner.
2. Reduce confusion. The only way to  the labels of a runner is modifying the config file.
3. Clear boundaries. Users cannot modify the labels if they cannot touch the runner, even if they are the admin of the Gitea instance.

### PR DESCRIPTION
**Pull request by sillyguodong**
close #24540

related:
- Protocol: https://gitea.com/gitea/actions-proto-def/pulls/9
- Runner side: https://gitea.com/gitea/act_runner/pulls/201

changes:
- Add column of `labels` to table `action_runner`, and combine the value of `agent_labels` and `custom_labels` column to `labels` column.
- Store `labels` when registering `act_runner`.
- Update `labels` when `act_runner` starting and calling `Declare`.
- Users cannot modify the `custom labels` in edit page any more.

other changes:
- Store `version` when registering `act_runner`.
- If runner is latest version, parse version from `Declare`. But older version runner still parse version from request header.

### CODE REVIEW

File: models/actions/runner.go [hunk 1]
```diff
@@ -43,10 +43,12 @@ type ActionRunner struct {
 	LastOnline timeutil.TimeStamp `xorm:"index"`
 	LastActive timeutil.TimeStamp `xorm:"index"`
 
-	// Store OS and Artch.
+	// Deprecated: Store OS and Artch.
 	AgentLabels []string
-	// Store custom labes use defined.
+	// Deprecated: Store custom labels use defined.
 	CustomLabels []string
```
Comment (wolfogre) [inline models/actions/runner.go]: I think we could just drop the two fields.

File: models/migrations/v1_20/v260.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+	}
+
+	// combine "agent labels" col and "custom labels" col to "labels" col.
+	err := x.Iterate(new(actions_model.ActionRunner), func(idx int, bean interface{}) error {
+		runner := bean.(*actions_model.ActionRunner)
+		runner.Labels = append(runner.Labels, runner.AgentLabels...)  //nolint
+		runner.Labels = append(runner.Labels, runner.CustomLabels...) //nolint
+		err := actions_model.UpdateRunner(db.DefaultContext, runner, "labels")
+		return err
+	})
```
Comment (wolfogre) [inline models/migrations/v1_20/v260.go]: Since `AgentLabels` and `CustomLabels` have been deleted, so maybe we should do this like:

1. `Sync`, it does't remove any columns
2. Combine the data  by a "more pure SQL way".
3. Drop `agent_labels` and `custom_labels`

File: routers/web/shared/actions/runners.go [hunk 3]
```diff
@@ -177,10 +175,11 @@ func RunnerDeletePost(ctx *context.Context, runnerID int64,
...
-	return labels
-}
+// unused, so comment out
+// func splitLabels(s string) []string {
+// 	labels := strings.Split(s, ",")
+// 	for i, v := range labels {
+// 		labels[i] = strings.TrimSpace(v)
+// 	}
+// 	return labels
+// }
```
Comment (wolfogre) [inline routers/web/shared/actions/runners.go]: Just remove unused codes, or why we use git as VCS? 😂

File: templates/shared/actions/runner_edit.tmpl [hunk 4]
```diff
@@ -16,9 +16,9 @@
 					<span>{{if .LastOnline}}{{TimeSinceUnix .LastOnline $.locale}}{{else}}{{$.locale.Tr "never"}}{{end}}</span>
 				</div>
 				<div class="field gt-dib gt-mr-4">
-					<label>{{.locale.Tr "actions.runners.agent_labels"}}</label>
+					<label>{{.locale.Tr "actions.runners.labels"}}</label>
```
Comment (wolfogre) [inline templates/shared/actions/runner_edit.tmpl:19]: I noticed there are some translations which need to be updated or removed.

`actions.runners.labels`, `actions.runners.custom_labels` and `actions.runners.custom_labels_helper`

File: models/migrations/migrations.go [hunk 5]
```diff
@@ -497,6 +497,8 @@ var migrations = []Migration{
 	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
 	// v259 -> 260
 	NewMigration("Convert scoped access tokens", v1_20.ConvertScopedAccessTokens),
+	// v260 -> v261
+	NewMigration("Add label column to action_run table, and combine labels", v1_20.AddLabelsToActRunner),
```
Comment (wolfogre) [inline models/migrations/migrations.go]: `1.20` got RC now, so it could be `1.21`

File: models/migrations/migrations.go [hunk 6]
```diff
@@ -497,6 +498,8 @@ var migrations = []Migration{
 	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
 	// v259 -> 260
 	NewMigration("Convert scoped access tokens", v1_20.ConvertScopedAccessTokens),
+	// v260 -> v261
```
Comment (wolfogre) [inline models/migrations/migrations.go:504]: `// Gitea 1.20.0 ends at v260`

Refer to previous.

File: routers/api/actions/runner/runner.go [hunk 7]
```diff
@@ -54,15 +54,23 @@ func (s *Service) Register(
 		return nil, errors.New("runner token has already been activated")
 	}
 
+	labels := req.Msg.Labels
+	// TODO: agent_labels should be removed from pb after Gitea 1.20 released.
+	// Old version runner's agent_labels slice is not empty and labels slice is empty.
+	// And due to compatibility with older versions, it is temporarily marked as Deprecated in pb, so use `//nolint` here.
+	if len(req.Msg.AgentLabels) > 0 && len(req.Msg.Labels) == 0 { //nolint
+		labels = req.Msg.AgentLabels //nolint
```
Comment (wolfogre) [inline routers/api/actions/runner/runner.go]:
```suggestion
	if len(req.Msg.AgentLabels) > 0 && len(req.Msg.Labels) == 0 { //nolint:staticcheck
		labels = req.Msg.AgentLabels //nolint:staticcheck
```

**Review by wolfogre — APPROVED** [review]
LGTM

File: models/migrations/v1_21/v260.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		return err
+	}
+
+	// combine "agent_labels" col and "custom_labels" col to "labels" col.
+	var runners []*ActionRunner
+	if err := sess.Table("action_runner").Select("id, agent_labels, custom_labels").Find(&runners); err != nil {
+		return err
+	}
+
+	for _, r := range runners {
```
Comment (lunny) [inline models/migrations/v1_21/v260.go]: How about
`UPDATE action_runner SET labels = agent_labels + custom_labels`
?

File: routers/web/shared/actions/runners.go [hunk 9]
```diff
@@ -126,9 +125,8 @@ func RunnerDetailsEditPost(ctx *context.Context, runnerID, ownerID, repoID int64
 
 	form := web.GetForm(ctx).(*forms.EditRunnerForm)
 	runner.Description = form.Description
-	runner.CustomLabels = splitLabels(form.CustomLabels)
 
-	err = actions_model.UpdateRunner(ctx, runner, "description", "custom_labels")
+	err = actions_model.UpdateRunner(ctx, runner, "description")
```
Comment (lunny) [inline routers/web/shared/actions/runners.go:129]: So then users cannot update labels from UI?

File: models/migrations/v1_21/v260.go (hunk 8 above)
Comment (wxiaoguang) [reply models/migrations/v1_21/v260.go]: I do not think this SQL works

The `labels` in model is ``Labels []string `xorm:"TEXT"` `` , XORM maintains the value (join/split). Unless the code breaks into XORM internal to imitate XORM's behavior, but it seems fragile.

File: routers/web/shared/actions/runners.go (hunk 9 above)
Comment (wxiaoguang) [reply routers/web/shared/actions/runners.go:129]: ~~It is described in the description. But I do not understand the new design either ...~~

~~The old design `AllLabes = AgentLables(from agent) + CustomLables(from user)` seems to work, no idea why there is a must to use the new design, only use one Lables field.~~

----

The full proposal is here:

* https://github.com/go-gitea/gitea/issues/24540

The reason is:

> So it will be convenient for users when they have a job with runs-on: centos, but there are no runners with the label. They can add centos as a custom label to let the runner pick up the jobs.
> Unfortunately, when the runner receives the job, it doesn't understand how to run it. To avoid failure, the act runner will use the default image to run jobs with unrecognized labels. This may surprise users.

File: models/migrations/v1_21/v260.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+		r.Labels = append(r.Labels, r.AgentLabels...)
+		r.Labels = append(r.Labels, r.CustomLabels...)
```
Comment (wxiaoguang) [inline models/migrations/v1_21/v260.go]: I think it should drop the CustomLables, according to the proposal, only agent labels can be used for a runner.

Comment (wolfogre) [reply models/migrations/v1_21/v260.go]: That's a good idea. And it means we don't have to combine the data and set a new column, just drop `CustomLabels` column and use `AgentLabels`.

Comment (sillyguodong) [reply models/migrations/v1_21/v260.go]: Since the `agent_labels` field is deprecated in pb and only `labels` is used, I want to keep Gitea be consistent with pb, so how about `UPDATE action_runner SET labels = agent_labels` and then drop the `agent_labels` and `custom_labels` columns?

Comment (sillyguodong) [reply models/migrations/v1_21/v260.go]: or use tag mapping?
```
Labels []string `xorm:"agent_labels"`
```

Comment (wolfogre) [reply models/migrations/v1_21/v260.go]: I don't really care the name of column, maybe
```go
Labels []string `xorm:"agent_labels"`
```
is enough, to make the migration simple.

Comment (wxiaoguang) [reply models/migrations/v1_21/v260.go]: Maybe we can still call this field as "AgentLabels", only drop the "CustomLables".

Indeed .... the labels are from agent, so AgentLabels is still a right name. And, maybe we would like to have custom labels support again in the future (for display & management purpose only and give them different color/styles to avoid confusing)

**Review by lunny — APPROVED** [review]

File: models/migrations/migrations.go [hunk 11]
```diff
@@ -497,6 +498,11 @@ var migrations = []Migration{
 	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
 	// v259 -> 260
 	NewMigration("Convert scoped access tokens", v1_20.ConvertScopedAccessTokens),
+
+	// Gitea 1.20.0 ends at 260
+
+	// v260 -> v261
+	NewMigration("Add label column to action_run table, and combine labels", v1_21.AddLabelsToActRunner),
```
Comment (wolfogre) [inline models/migrations/migrations.go]: The description of the migration should be updated now.

File: models/migrations/v1_21/v260.go [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+
+package v1_21 //nolint
+
+import (
+	"code.gitea.io/gitea/models/migrations/base"
+
+	"xorm.io/xorm"
+)
+
+func AddLabelsToActRunner(x *xorm.Engine) error {
```
Comment (wolfogre) [inline models/migrations/v1_21/v260.go]: And the function name.

File: models/migrations/migrations.go [hunk 13]
```diff
@@ -497,6 +498,11 @@ var migrations = []Migration{
 	NewMigration("Add PinOrder Column", v1_20.AddPinOrderToIssue),
 	// v259 -> 260
 	NewMigration("Convert scoped access tokens", v1_20.ConvertScopedAccessTokens),
+
+	// Gitea 1.21.0 ends at 260
```
Comment (wolfogre) [inline models/migrations/migrations.go:502]: Wait a minute, why did you change it to `1.21.0` in 6cc62e9b6d5c880606813ab90d8db720bac7a6b7 ? Please be more careful.

Comment (wolfogre) [reply models/migrations/migrations.go:502]: Alright, I see #25243.

Comment (sillyguodong) [reply models/migrations/migrations.go:502]: will be more careful next time...

**Comment by ExplodingDragon** [comment]
Will this patch be backported to 1.20?

**Comment by lunny** [comment]
> Will this patch be backported to 1.20?

It will not
